.github/workflows/docker.yaml

```diff
@@ -0,0 +1,84 @@
+name: Docker
+
+on:
+  push:
+  pull_request:
+    branches: [ master ]
+
+jobs:
+  build:
+    name: Build and push Docker image
+    runs-on: ubuntu-latest
+    env:
+      IMAGE: ghcr.io/${{ github.repository }}
+
+    permissions:
+      packages: write
+      contents: read
+
+    steps:
+      - name: Checkout the code
+        uses: actions/checkout@v2
+
+      - name: Docker meta
+        id: meta
+        uses: docker/metadata-action@v3
+        with:
+          images: "${{ env.IMAGE }}"
+          bake-target: docker-metadata-action
+          tags: |
+            type=ref,event=branch
+            type=semver,pattern={{version}}
+            type=semver,pattern={{major}}.{{minor}}
+            type=semver,pattern={{major}}
+            type=sha
```

Comment:
I'm struggling to get my head round this a bit. is the first match in the list used? or what?

Reply:
Those are the rules for generating the list of tags. It's documented here: https://github.com/docker/metadata-action#tags-input

`type=ref,event=branch`: a push to a branch named `foo` creates a tag named `foo`
`type=semver,pattern={{version}}`: a `v1.2.3` git tag creates a `1.2.3` docker tag. I think it also sets the `latest` tag?
`type=semver,pattern={{major}}.{{minor}}`: a `v1.2.3` git tag creates a `1.2` docker tag
`type=semver,pattern={{major}}`: a `v1.2.3` git tag creates a `1` docker tag
`type=sha`: any commit generates a `sha-aabbccd`, where `aabbccd` are the 7 first digits of the commit hash

The rules below are the same, but with a `-debug` suffix (e.g. `:1.2.3-debug`)

Reply:
So the latest push on this PR branch (`sandhose/docker-improvements` / ab3c3d8e460e3d34e4bd53d813b8990abc260516) pushed the following tags:

- `sha-ab3c3d8-debug` and `sandhose-docker-improvements-debug` for the debug variant
- `sha-ab3c3d8` and `sandhose-docker-improvements` for the regular image

See https://github.com/matrix-org/rageshake/pkgs/container/rageshake

Reply:
oh so they are *all* applied. Sorry, I was having a bit of a brainfail.